core/src/main/c/share/dedup.cpp

```diff
@@ -0,0 +1,644 @@
+    int64_t reserved3;
+    char null_value[32];
+};
+#pragma pack(pop)
+
+template<typename T>
+struct dedup_column_t : dedup_column {
+    [[nodiscard]] inline const T *get_column_data() const {
+        return (T *) dedup_column::column_data;
+    };
+
+    [[nodiscard]] inline const T *get_o3_data() const {
+        return (T *) dedup_column::o3_data;
+    };
+
+    [[nodiscard]] inline const T &get_null_value() const {
+        return (T &) dedup_column::null_value;
+    };
+};
+
+struct int256 {
+    __int128 lo;
+    __int128 hi;
+};
+
+inline bool operator>(const int256 &a, const int256 &b) {
+    return a.hi > b.hi || a.lo > b.lo;
+}
+
+inline bool operator<(const int256 &a, const int256 &b) {
+    return a.hi < b.hi || a.lo < b.lo;
+}
+
+template<typename lambda_diff>
+inline int64_t branch_free_search(int64_t searchIndex, const index_t *array, const int64_t count, lambda_diff compare) {
+    const index_t *base = array;
+    int64_t n = count;
+    while (n > 1) {
+        int64_t half = n / 2;
+        MM_PREFETCH_T0(base + half / 2);
+        MM_PREFETCH_T0(base + half + half / 2);
+        auto diff = compare(searchIndex, base[half].i);
+        base = (diff > 0) ? base + half : base;
+        n -= half;
+    }
+    if (compare(searchIndex, base[0].i) == 0) {
+        return base - array;
+    }
+    if (base - array + 1 < count && compare(searchIndex, base[1].i) == 0) {
+        return base - array + 1;
+    }
+    return -1;
+}
+
+template<typename lambda_diff>
+int64_t merge_dedup_long_index_int_keys(
+        const uint64_t *src,
+        const int64_t src_lo,
+        const int64_t src_hi_incl,
+        index_t *const index,
+        const int64_t index_lo,
+        const int64_t index_hi_incl,
+        index_t *dest_index,
+        const lambda_diff compare
+) {
+    int64_t src_pos = src_lo;
+    int64_t index_pos = index_lo;
+    index_t *dest = dest_index;
+
+    bit_vector used_indexes = {};
+    while (src_pos <= src_hi_incl && index_pos <= index_hi_incl) {
+        // Perform normal merge until the timestamp matches.
+        if (src[src_pos] < index[index_pos].ts) {
+            dest[0].ts = src[src_pos];
+            dest[0].i = src_pos | (1ull << 63);
+            dest++;
+            src_pos++;
+        } else if (src[src_pos] > index[index_pos].ts) {
+            *dest = index[index_pos];
+            dest++;
+            index_pos++;
+        } else {
+            // index_ts == src_ts
+            const uint64_t conflict_ts = src[src_pos];
+            index_t *const conflict_index_start = &index[index_pos];
+
+            // Find end of the conflict in index
+            const auto conflict_end_pos = index_pos +
+                                          branch_free_search_lower<index_t>(&index[index_pos],
+                                                                            index_hi_incl - index_pos + 1,
+                                                                            conflict_ts + 1);
```

Comment:
I would suggest to pass `index_t {conflict_ts + 1, 0}` here and remove V template parameter from `branch_free_search_lower`